File: hospital_readmission/modules/evaluator.py
import numpy as np
import pandas as pd
from sklearn.metrics import (
    classification_report, confusion_matrix,
    roc_curve, roc_auc_score,
    precision_score, recall_score,
    f1_score, accuracy_score
)
from modules.model import _model_state, MODELS

def get_available_models():
    return [
        {"key": k, "name": MODELS[k]["name"]}
        for k in _model_state["trained_models"]
    ]

def evaluate_model(model_key):
    if model_key not in _model_state["trained_models"]:
        return {"error": f"Model '{model_key}' not trained yet"}

    clf     = _model_state["trained_models"][model_key]
    X_test  = _model_state["X_test"]
    y_test  = _model_state["y_test"]
    X_train = _model_state["X_train"]
    y_train = _model_state["y_train"]
    features = _model_state["feature_names"]

    y_pred       = clf.predict(X_test)
    y_pred_proba = clf.predict_proba(X_test)[:, 1] \
                   if hasattr(clf, "predict_proba") else None

    # ── Confusion matrix ──
    cm = confusion_matrix(y_test, y_pred)
    tn, fp, fn, tp = cm.ravel()

    # ── Classification report ──
    report = classification_report(
        y_test, y_pred,
        target_names=["Not readmitted", "Readmitted"],
        output_dict=True
    )

    # ── ROC curve ──
    fpr_arr, tpr_arr, _ = roc_curve(y_test, y_pred_proba) \
        if y_pred_proba is not None else ([], [], [])
    auc = round(float(roc_auc_score(y_test, y_pred_proba)), 4) \
          if y_pred_proba is not None else None

    step = max(1, len(fpr_arr) // 60)
    fpr = [round(float(v), 4) for v in fpr_arr[::step]]
    tpr = [round(float(v), 4) for v in tpr_arr[::step]]

    # ── Feature importance ──
    importance = []
    if hasattr(clf, "feature_importances_"):
        imp = pd.Series(clf.feature_importances_, index=features)
        imp = imp.sort_values(ascending=False).head(15)
        importance = [
            {"feature": f, "importance": round(float(v), 4)}
            for f, v in imp.items()
        ]
    elif hasattr(clf, "coef_"):
        imp = pd.Series(np.abs(clf.coef_[0]), index=features)
        imp = imp.sort_values(ascending=False).head(15)
        importance = [
            {"feature": f, "importance": round(float(v), 4)}
            for f, v in imp.items()
        ]

    # ── Overfitting check ──
    train_acc = round(float(accuracy_score(y_train, clf.predict(X_train))), 4)
    test_acc  = round(float(accuracy_score(y_test, y_pred)), 4)
    overfit_gap = round(float(train_acc - test_acc), 4)

    return {
        "model_key":  model_key,
        "model_name": MODELS[model_key]["name"],
        "confusion_matrix": {
            "tn": int(tn), "fp": int(fp),
            "fn": int(fn), "tp": int(tp)
        },
        "metrics": {
            "accuracy":  round(float(accuracy_score(y_test, y_pred)), 4),
            "f1":        round(float(f1_score(y_test, y_pred, zero_division=0)), 4),
            "precision": round(float(precision_score(y_test, y_pred, zero_division=0)), 4),
            "recall":    round(float(recall_score(y_test, y_pred, zero_division=0)), 4),
            "roc_auc":   auc,
            "train_acc": train_acc,
            "test_acc":  test_acc,
            "overfit_gap": overfit_gap
        },
        "classification_report": {
            "not_readmitted": {
                "precision": round(report["Not readmitted"]["precision"], 4),
                "recall":    round(report["Not readmitted"]["recall"], 4),
                "f1":        round(report["Not readmitted"]["f1-score"], 4),
                "support":   int(report["Not readmitted"]["support"])
            },
            "readmitted": {
                "precision": round(report["Readmitted"]["precision"], 4),
                "recall":    round(report["Readmitted"]["recall"], 4),
                "f1":        round(report["Readmitted"]["f1-score"], 4),
                "support":   int(report["Readmitted"]["support"])
            }
        },
        "roc_curve": {"fpr": fpr, "tpr": tpr, "auc": auc},
        "feature_importance": importance,
        "overfit": {
            "train_acc":    train_acc,
            "test_acc":     test_acc,
            "gap":          overfit_gap,
            "status": "Overfitting" if overfit_gap > 0.1
                      else "Slight overfit" if overfit_gap > 0.05
                      else "Good fit"
        }
    }

File: hospital_readmission/static/js/evaluator.js
async function loadEvalModels() {
  const models = await fetch('/api/eval/models').then(r => r.json());
  const sel = document.getElementById('eval-model-select');
  sel.innerHTML = models.length
    ? models.map(m =>
        `<option value="${m.key}">${m.name}</option>`
      ).join('')
    : '<option value="">No models trained yet</option>';
}

async function runEvaluation() {
  const key = document.getElementById('eval-model-select').value;
  if (!key) { alert('No model selected'); return; }

  const btn = event.target;
  btn.textContent = 'Evaluating...';
  btn.disabled = true;

  const d = await fetch(`/api/eval/${key}`).then(r => r.json());
  if (d.error) {
    alert(d.error);
    btn.textContent = 'Evaluate';
    btn.disabled = false;
    return;
  }

  renderMetricCards(d.metrics);
  renderConfusionMatrix(d.confusion_matrix);
  renderROCCurve(d.roc_curve);
  renderClassReport(d.classification_report);
  renderOverfit(d.overfit);
  if (d.feature_importance.length) renderFeatureImportance(d.feature_importance);
  else document.getElementById('feat-imp-card').style.display = 'none';

  document.getElementById('eval-results').style.display = 'block';
  btn.textContent = 'Evaluate';
  btn.disabled = false;
}

function renderMetricCards(m) {
  document.getElementById('ev-acc').textContent    = (m.accuracy * 100).toFixed(1) + '%';
  document.getElementById('ev-f1').textContent     = m.f1;
  document.getElementById('ev-auc').textContent    = m.roc_auc || '—';
  document.getElementById('ev-recall').textContent = m.recall;
}

function renderConfusionMatrix(cm) {
  const total = cm.tn + cm.fp + cm.fn + cm.tp;
  const cells = [
    { label: 'True Negative',  val: cm.tn, bg: '#EAF3DE', color: '#3B6D11',
      desc: 'Correctly predicted NOT readmitted' },
    { label: 'False Positive', val: cm.fp, bg: '#FAEEDA', color: '#854F0B',
      desc: 'Predicted readmitted — actually not' },
    { label: 'False Negative', val: cm.fn, bg: '#FCEBEB', color: '#A32D2D',
      desc: 'Predicted not readmitted — actually was' },
    { label: 'True Positive',  val: cm.tp, bg: '#E6F1FB', color: '#185FA5',
      desc: 'Correctly predicted readmitted' }
  ];

  document.getElementById('cm-grid').innerHTML = cells.map(c => `
    <div style="background:${c.bg};border-radius:8px;padding:14px;text-align:center">
      <div style="font-size:11px;font-weight:600;color:${c.color};
                  text-transform:uppercase;margin-bottom:6px">${c.label}</div>
      <div style="font-size:28px;font-weight:700;color:${c.color}">${c.val.toLocaleString()}</div>
      <div style="font-size:11px;color:${c.color};opacity:0.8;margin-top:4px">
        ${((c.val / total) * 100).toFixed(1)}%
      </div>
      <div style="font-size:10px;color:${c.color};opacity:0.7;margin-top:4px">
        ${c.desc}
      </div>
    </div>`).join('');

  document.getElementById('cm-legend').innerHTML =
    `Total predictions: <strong>${total.toLocaleString()}</strong> &nbsp;|&nbsp;
     Correct: <strong>${(cm.tn + cm.tp).toLocaleString()}</strong> &nbsp;|&nbsp;
     Wrong: <strong>${(cm.fp + cm.fn).toLocaleString()}</strong>`;
}

function renderROCCurve(roc) {
  if (window._rocChart) window._rocChart.destroy();
  if (!roc.fpr.length) return;

  window._rocChart = new Chart(document.getElementById('roc-canvas'), {
    type: 'line',
    data: {
      labels: roc.fpr,
      datasets: [
        {
          label: 'ROC Curve',
          data: roc.tpr,
          borderColor: '#185FA5',
          backgroundColor: 'rgba(24,95,165,0.08)',
          borderWidth: 2,
          pointRadius: 0,
          fill: true,
          tension: 0.3
        },
        {
          label: 'Random classifier',
          data: roc.fpr,
          borderColor: '#D3D1C7',
          borderWidth: 1,
          borderDash: [5, 5],
          pointRadius: 0,
          fill: false
        }
      ]
    },
    options: {
      responsive: true,
      maintainAspectRatio: false,
      plugins: { legend: { display: false } },
      scales: {
        x: { title: { display: true, text: 'False Positive Rate', font:{size:11} },
             ticks: { font: { size: 10 } }, min: 0, max: 1 },
        y: { title: { display: true, text: 'True Positive Rate', font:{size:11} },
             ticks: { font: { size: 10 } }, min: 0, max: 1 }
      }
    }
  });

  document.getElementById('roc-auc-label').textContent =
    `AUC = ${roc.auc}  (closer to 1.0 = better)`;
}

function renderClassReport(report) {
  const tbody = document.getElementById('clf-report-body');
  const rows  = [
    { label: 'Not readmitted (0)', ...report.not_readmitted },
    { label: 'Readmitted (1)',     ...report.readmitted }
  ];
  tbody.innerHTML = rows.map(r => `
    <tr>
      <td><strong>${r.label}</strong></td>
      <td>${r.precision}</td>
      <td>${r.recall}</td>
      <td>${r.f1}</td>
      <td>${r.support.toLocaleString()}</td>
    </tr>`).join('');
}

function renderOverfit(ov) {
  const color  = ov.status === 'Good fit'      ? '#3B6D11'
               : ov.status === 'Slight overfit' ? '#854F0B'
               : '#A32D2D';
  const bg     = ov.status === 'Good fit'      ? '#EAF3DE'
               : ov.status === 'Slight overfit' ? '#FAEEDA'
               : '#FCEBEB';

  document.getElementById('overfit-display').innerHTML = `
    <div style="display:flex;align-items:center;gap:16px;flex-wrap:wrap">
      <div class="stat-item">
        <div class="lbl">Train accuracy</div>
        <div class="val">${(ov.train_acc * 100).toFixed(1)}%</div>
      </div>
      <div style="font-size:20px;color:#aaa">→</div>
      <div class="stat-item">
        <div class="lbl">Test accuracy</div>
        <div class="val">${(ov.test_acc * 100).toFixed(1)}%</div>
      </div>
      <div style="font-size:20px;color:#aaa">→</div>
      <div class="stat-item">
        <div class="lbl">Gap</div>
        <div class="val">${(ov.gap * 100).toFixed(1)}%</div>
      </div>
      <div style="background:${bg};color:${color};padding:8px 18px;
           border-radius:20px;font-size:13px;font-weight:600">
        ${ov.status}
      </div>
    </div>
    <div style="font-size:12px;color:#888;margin-top:10px">
      Gap > 10% = Overfitting &nbsp;|&nbsp;
      Gap 5–10% = Slight overfit &nbsp;|&nbsp;
      Gap &lt; 5% = Good fit
    </div>`;
}

function renderFeatureImportance(importance) {
  if (window._impChart) window._impChart.destroy();

  const labels = importance.map(i => i.feature);
  const values = importance.map(i => i.importance);
  const maxVal = Math.max(...values);

  window._impChart = new Chart(document.getElementById('feat-imp-canvas'), {
    type: 'bar',
    data: {
      labels,
      datasets: [{
        label: 'Importance',
        data: values,
        backgroundColor: values.map((v, i) =>
          i === 0 ? '#0C447C' : i < 3 ? '#185FA5' :
          i < 7  ? '#378ADD' : '#85B7EB'
        ),
        borderRadius: 4,
        borderWidth: 0
      }]
    },
    options: {
      indexAxis: 'y',
      responsive: true,
      maintainAspectRatio: false,
      plugins: { legend: { display: false } },
      scales: {
        x: { grid: { display: false },
             ticks: { font: { size: 11 } } },
        y: { grid: { display: false },
             ticks: { font: { size: 11 } } }
      }
    }
  });
}